src/Microsoft.Health.Fhir.Api/Features/Binders/DictionaryExpansionConfigurationProvider.cs

```diff
@@ -0,0 +1,78 @@
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using EnsureThat;
+using Microsoft.Extensions.Configuration;
+using Newtonsoft.Json;
+
+namespace Microsoft.Health.Fhir.Api.Features.Binders;
+
+/// <summary>
+/// Detects JSON dictionaries in variables and converts them to <see cref="IConfiguration"/> values.
+/// This allows the built-in binder to hydrate <see cref="Dictionary{TKey,TValue}"/> typed configuration properties.
+/// </summary>
+/// <remarks>
+/// This configuration provider ignores other variables if not in JSON format.
+/// </remarks>
+public class DictionaryExpansionConfigurationProvider : ConfigurationProvider
+{
+    private readonly IConfigurationProvider _configurationProvider;
+
+    /// <summary>
+    /// Creates a <see cref="DictionaryExpansionConfigurationProvider"/> with a customer <see cref="IConfigurationProvider"/>.
+    /// </summary>
+    /// <param name="configurationProvider">Custom configuration provider.</param>
+    /// <remarks>
+    /// Constructor only used for testing purposes.
+    /// </remarks>
+    public DictionaryExpansionConfigurationProvider(IConfigurationProvider configurationProvider)
+    {
+        EnsureArg.IsNotNull(configurationProvider, nameof(configurationProvider));
+
+        _configurationProvider = configurationProvider;
+        Data = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+    }
+
+    public override void Load()
+    {
+        base.Load();
+
+        _configurationProvider.Load();
+
+        IEnumerable<string> keys = _configurationProvider.GetChildKeys(Array.Empty<string>(), null);
+
+        var data = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+        EnumerateKeys(keys, data);
+
+        Data = data;
+    }
+
+    private void EnumerateKeys(IEnumerable<string> keys, Dictionary<string, string> data, string path = null)
+    {
+        foreach (string keyName in keys.Distinct())
+        {
+            var keyPath = path != null ? $"{path}:{keyName}" : keyName;
+
+            _configurationProvider.TryGet(keyPath, out string environmentVariableValue);
+
+            if (!string.IsNullOrEmpty(environmentVariableValue) && environmentVariableValue.Trim().StartsWith("{", StringComparison.Ordinal) == true)
+            {
+                Dictionary<string, string> asDictionary = JsonConvert.DeserializeObject<Dictionary<string, string>>(environmentVariableValue);
+
+                foreach (KeyValuePair<string, string> kvp in asDictionary!)
+                {
+                    data.Add($"{keyPath}:{kvp.Key}", kvp.Value);
+                }
+            }
+
+            IEnumerable<string> innerKeys = _configurationProvider.GetChildKeys(Array.Empty<string>(), keyPath);
+            EnumerateKeys(innerKeys, data, keyPath);
```

Comment:
While testing it looks like we need to build the keypath to find the nested values

Reply:
Just a thought, is this something that can be covered in tests?

Reply:
@fhibf added more below. Also now that the MockConfigurationProvider inherits ConfigurationProvider it does drill into this